File: valid-sudoku/valid-sudoku-mustafasegf.go
package validsudoku

func isValidSudoku(board [][]byte) bool {
	row := make([]map[byte]bool, 9)
	col := make([]map[byte]bool, 9)
	box := make([][]map[byte]bool, 3)

	for i := range row {
		row[i] = make(map[byte]bool)
	}

	for i := range col {
		col[i] = make(map[byte]bool)
	}

	for j := range box {
		box[j] = make([]map[byte]bool, 3)
		for i := range box[j] {
			box[j][i] = make(map[byte]bool)
		}
	}

	for i, arr := range board {
		for j, c := range arr {
			if c == 46 {
				continue
			}
			if _, ok := row[j][c]; ok {
				return false
			}
			row[j][c] = true

			if _, ok := col[i][c]; ok {
				return false
			}
			col[i][c] = true

			if _, ok := box[i/3][j/3][c]; ok {
				return false
			}
			box[i/3][j/3][c] = true

		}
	}

	return true
}

File: valid-sudoku/valid-sudoku-mustafasegf.rs
use std::collections::HashSet;

impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let mut rows = [0; 9];
        let mut cols = [0; 9];
        let mut grid = [0; 9];

        for row in 0..9 {
            for col in 0..9 {
                if let Some(num) = board[row][col].to_digit(10) {
                    let grid_idx = 3 * (row / 3) + (col / 3);
                    let num = 1 << num;

                    if (rows[row] & num) != 0
                        || (cols[col] & num) != 0
                        || (grid[grid_idx] & num) != 0
                    {
                        return false;
                    }

                    rows[row] |= num;
                    cols[col] |= num;
                    grid[grid_idx] |= num;
                }
            }
        }

        return true;
    }
}
